client/web/web.go

```diff
@@ -206,13 +237,130 @@ func (s *Server) serveLoginAPI(w http.ResponseWriter, r *http.Request) {
 	return
 }
 
+var (
+	errNoSession         = errors.New("no-browser-session")
+	errNotUsingTailscale = errors.New("not-using-tailscale")
+	errTaggedSource      = errors.New("tagged-source")
+	errNotOwner          = errors.New("not-owner")
+)
+
+// getTailscaleBrowserSession retrieves the browser session associated with
+// the request, if one exists.
+//
+// An error is returned in any of the following cases:
+//
+//   - (errNotUsingTailscale) The request was not made over tailscale.
+//
+//   - (errNoSession) The request does not have a session.
+//
+//   - (errTaggedSource) The source is a tagged node. Users must use their
+//     own user-owned devices to manage other nodes' web clients.
+//
+//   - (errNotOwner) The source is not the owner of this client (if the
+//     client is user-owned). Only the owner is allowed to manage the
+//     node via the web client.
+//
+// If no error is returned, the browserSession is always non-nil.
+// getTailscaleBrowserSession does not check whether the session has been
+// authorized by the user. Callers can use browserSession.isAuthorized.
+func (s *Server) getTailscaleBrowserSession(r *http.Request) (*browserSession, error) {
+	whoIs, err := s.lc.WhoIs(r.Context(), r.RemoteAddr)
+	switch {
+	case err != nil || whoIs.Node == nil:
+		return nil, errNotUsingTailscale
```

Comment:
there's no harm in having this, but should this ever happen?  Wouldn't this only be exposed as part of the full client that is only available on the tailscale interface?

Maybe what I'm really suggesting is that this is a more fundamental error state (whois lookup not working), but shouldn't actually be related to them not using tailscale.  I think the error is just misnamed.

Reply:
Oh yes, I see what you're saying. The docs for the whois response in `apitype.go` specify:
```
// WhoIsResponse is the JSON type returned by tailscaled debug server's /whois?ip=$IP handler.
// In successful whois responses, Node and UserProfile are never nil.
type WhoIsResponse struct {
```
So yeah, it looks like if err is nil, `whoIs.Node` should always be filled. So I think it's safe for me to remove this check here.

Reply:
Same with my `UserProfile == nil` check in the next case.